0003-interledger-protocol/0003-interledger-protocol.md

```diff
@@ -279,24 +277,25 @@ Inspired by [HTTP Status Codes](https://tools.ietf.org/html/rfc2616#section-10),
 
-| Code | Name | Description |
-|---|---|---|
-| **F00** | **Bad Request** | Generic sender error. |
-| **F01** | **Invalid Packet** | The ILP packet was syntactically invalid. |
-| **F02** | **Unreachable** | There was no way to forward the payment, because the destination ILP address was wrong or the connector does not have a route to the destination. |
-| **F03** | **Invalid Amount** | The amount is invalid, for example it contains more digits of precision than are available on the destination ledger or the amount is greater than the total amount of the given asset in existence. |
-| **F04** | **Insufficient Destination Amount** | The receiver deemed the amount insufficient, for example you tried to pay a $100 invoice with $10. |
-| **F05** | **Wrong Condition** | The receiver generated a different condition and cannot fulfill the payment. |
-| **F06** | **Unexpected Payment** | The receiver was not expecting a payment like this (the memo and destination address don't make sense in that combination, for example if the receiver does not understand the transport protocol used) |
-| **F07** | **Cannot Receive** | The receiver is unable to accept this payment due to a constraint. For example, the payment would put the receiver above its maximum account balance. |
-| **F99** | **Application Error** | Reserved for application layer protocols. Applications MAY use names other than `Application Error`. |
+| Code | Name | Description | Data Fields |
+|---|---|---|---|
+| **F00** | **Bad Request** | Generic sender error. | |
+| **F01** | **Invalid Packet** | The ILP packet was syntactically invalid. | |
+| **F02** | **Unreachable** | There was no way to forward the payment, because the destination ILP address was wrong or the connector does not have a route to the destination. | |
+| **F03** | **Invalid Amount** | The amount is invalid, for example it contains more digits of precision than are available on the destination ledger or the amount is greater than the total amount of the given asset in existence. | |
+| **F04** | **Insufficient Destination Amount** | The receiver deemed the amount insufficient, for example you tried to pay a $100 invoice with $10. | |
+| **F05** | **Wrong Condition** | The receiver generated a different condition and cannot fulfill the payment. | |
+| **F06** | **Unexpected Payment** | The receiver was not expecting a payment like this (the memo and destination address don't make sense in that combination, for example if the receiver does not understand the transport protocol used) | |
+| **F07** | **Cannot Receive** | The receiver is unable to accept this payment due to a constraint. For example, the payment would put the receiver above its maximum account balance. | |
+| **F08** | **Payment Too Large** | The payment is larger than the maximum size supported by a connector, ledger or the receiver. | `SourceAmount` (UInt64), `SourceAmountLimit` (UInt64) |
```

Comment:
According to OP, the data fields `SourceAmount` and `SourceAmountLimit` should help the sender adjust the payment amount. So the procedure the sender would adjust the payment size is:

1. the sender calculates `OverpayRatio` = `SourceAmount` / `SourceAmountLimit`
2. the sender divides the amount of the failed payment by `OverpayRatio` to obtain the new payment amount, which should be small enough to avoid a `Payment Too Large` error.

Is that correct? If so, how about replacing the data fields `SourceAmount` and `SourceAmountLimit` directly with `OverpayRatio`? The sender is likely not familiar with the unit of the connector's `SourceAmount`, so absolute integers are meaningless to the sender. Furthermore, it would save same bytes (one 32 bit float as opposed to two 64 bit integer).

Alternatively, if we stick with `SourceAmount` and `SourceAmountLimit`, each connector that passes on a `Payment Too Large` error could adjust the `SourceAmount` and `SourceAmountLimit` under consideration of it's own exchange rate (like with a "Quote by Destination Amount"). By doing so, the sender would obtain amounts in his own ledger units and adjusting the payment size would be trivial.

Reply:
That's correct.

> how about replacing the data fields SourceAmount and SourceAmountLimit directly with OverpayRatio?

Fair question. I guess it mostly comes down to whether we want to include a float (losing some precision -- also it should be a 64 bit float at least) or the two integers. I don't feel too strongly about this.

Either way, I feel strongly that the connectors along the way should not try to modify the error. That would add another feature that every connector must implement (it would get seriously messed up if some connectors adjusted the values but not all), rather than limiting the feature to the party that has the payment size limit.

Reply:
> I guess it mostly comes down to whether we want to include a float (losing some precision -- also it should be a 64 bit float at least) or the two integers.

Since exchange rates fluctuate (and liquidity curves don't have to be linear, at least for the moment), `OverpayRatio`/`SourceAmount` is anyway only an estimate. Therefore, I am not too much concerned with the loss of precision due to the floating point operation. Most likely a sender would subtract a bit more from the payment amount than what is indicated by `OverpayRatio` to avoid another `Payment Too Large` error.

Reply:
The main arguments for using two integers from my perspective:

* Floats allow special values like infinity, negative zero, NaN and subnormal numbers which could be used to trigger bugs or strange behavior in implementations.
* We already have 64-bit unsigned ints used for amounts in various places whereas using a float would introduce a new type that people have to implement in their parsers.
* Floats introduce some amount of precision error whereas the actual amount and max amount values are (usually) going to be actual values that tripped some *if* statement somewhere. If nothing else, that seems good for debugging.

Taken together I think we should use two integers.

One thing I don't understand is that it's called `SourceAmount` and `SourceAmountLimit`. That seems to indicate that it's the connector receiving the overpayment that is complaining. But wouldn't limits usually be on trustlines/channels? And if so, wouldn't the connector sending that transfer notice the fact that the payment would be over the limit (and be the one to trigger the error?)

I also don't like `SourceAmount` since we usually mean the source amount of the overall ILP path, not the *locally* incoming amount.

So I'd propose to use instead: `LocalAmount` and `LocalAmountLimit` which is neutral in terms of incoming vs outgoing and gives you a hint that this amount is in the complainant's local unit which may be different than the units of the source and destination ledgers.

Another problem is that the amount may be larger than UINT64_MAX. This could be an argument for using floats, but a) I think that this would be very rare and 2) all that would happen is that you would cap the `LocalAmount` at UINT64_MAX which would in the worst case trigger an extra overpayment or two. Plus, if there is a connector where this happens a lot and they don't mind the extra complexity, they could also adjust `LocalAmountLimit` downwards to ensure faster convergence.